script/core/diagnostics/missing-fields.lua

```diff
@@ -15,61 +15,74 @@ return function (uri, callback)
     guide.eachSourceType(state.ast, 'table', function (src)
         await.delay()
 
+        vm.removeNode(src) -- the node is not updated correctly, reason still unknown
         local defs = vm.getDefs(src)
+        local sortedDefs = {}
         for _, def in ipairs(defs) do
-            if def.type == 'doc.class' and def.bindSource then
-                if guide.isInRange(def.bindSource, src.start) then
+            if def.type == 'doc.class' then
+                if def.bindSource and guide.isInRange(def.bindSource, src.start) then
                     return
                 end
+                local className = def.class[1]
+                if not sortedDefs[className] then
+                    sortedDefs[className] = {}
+                end
+                local samedefs = sortedDefs[className]
+                samedefs[#samedefs+1] = def
             end
             if def.type == 'doc.type.array'
             or def.type == 'doc.type.table' then
                 return
             end
         end
-        local warnings = {}
-        for _, def in ipairs(defs) do
-            if def.type == 'doc.class' then
-                if not def.fields then
-                    return
-                end
 
-                local requiresKeys = {}
-                for _, field in ipairs(def.fields) do
-                    if  not field.optional
-                    and not vm.compileNode(field):isNullable() then
-                        local key = vm.getKeyName(field)
-                        if key and not requiresKeys[key] then
-                            requiresKeys[key] = true
-                            requiresKeys[#requiresKeys+1] = key
-                        end
-                    end
+        local warnings = {}
+        for _, samedefs in pairs(sortedDefs) do
+            local missedKeys = {}
+            local className
+            for _, def in ipairs(samedefs) do
+                className = def.class[1]
```

Comment:
I see that `sortedDefs{}` are already indexed by `className`
=> `sortedDefs = { [className] = samedefs[] }`

Is there any reason no getting `className` directly as forloop variable in the outer `for pairs` loop? 😕 
```suggestion
        for className, samedefs in pairs(sortedDefs) do
            local missedKeys = {}
            for _, def in ipairs(samedefs) do
```